File: packages/dirigent-server/frontend/src/lib/schemas.ts
/**
 * The schema resources, as this bundle reads them.
 *
 * THE FIELD NAMES ARE THE WIRE'S. Every interface here mirrors a pydantic model in
 * `dirigent_client.schemas.schemas` member for member.
 *
 * A SCHEMA IS LOCALLY AUTHORED. What an instance holds is a JSON Schema somebody wrote to
 * say what a payload should look like -- never fetched or introspected from anywhere. Its
 * code, name and description are the schema's own `$id`, `title` and `description`, settled
 * when it is stored, so what this bundle reads back is a portable schema rather than a
 * wrapper around one.
 */

import { apiJson, apiSend, type JsonMap, type Page } from '@/lib/api'
import { PAGE } from '@/lib/paging'

/** A schema as every response shows it: the identity quartet and the schema body. `SchemaOut`. */
export interface SchemaOut {
    id: string
    /** The key this schema is addressed by, and the name `requires.schemas` reaches it under. */
    code: string
    /** What to call it on screen, when the schema gave itself a `title`. */
    name: string | null
    /** Long-form, from the schema's own `description`. */
    description: string | null
    /** The JSON Schema itself. */
    body: JsonMap
    created_at: string
    updated_at: string
}

/** Read one page of schemas, in code order. */
export function readSchemas(after: string | null = null): Promise<Page<SchemaOut>> {
    const query = new URLSearchParams({ limit: String(PAGE) })
    if (after !== null) query.set('after', after)
    return apiJson<Page<SchemaOut>>(`/schemas?${query.toString()}`)
}

/** Store a JSON Schema, letting the server take its identity from the schema's own keywords. */
export function createSchema(body: JsonMap, code: string | null): Promise<SchemaOut> {
    const payload: JsonMap = { body }
    if (code !== null && code !== '') payload.code = code
    return apiSend<SchemaOut>('/schemas', 'POST', payload)
}

/** Remove a schema. */
export function deleteSchema(code: string): Promise<void> {
    return apiSend<void>(`/schemas/${encodeURIComponent(code)}`, 'DELETE', {})
}

/**
 * What the schemas listing says when it has nothing in it.
 *
 * A schema is authored, not discovered, so an empty instance is one nobody has written a
 * shape for yet -- which is what the empty state says, rather than a bare "none".
 */
export function schemasNote(rows: readonly SchemaOut[]): string | null {
    if (rows.length === 0) return null
    return `${String(rows.length)} ${rows.length === 1 ? 'schema' : 'schemas'}`
}

File: packages/dirigent-client/src/dirigent_client/resources/schemas.py
"""Schemas: named JSON Schemas an instance holds, addressable by code and referenced by name."""

from dirigent_client.resources.base import Resource, query, request_body
from dirigent_client.schemas import Page, SchemaIn, SchemaOut, SchemaUpdate
from dirigent_common import JsonMap


class Schemas(Resource):
    """Create, read, edit, and remove the JSON Schemas an instance holds."""

    async def list(self, *, after: str | None = None, limit: int | None = None) -> Page[SchemaOut]:
        """List every stored schema."""
        return await self._many(SchemaOut, "GET", "/schemas", params=query(after=after, limit=limit))

    async def get(self, code: str) -> SchemaOut:
        """Read one schema by its code."""
        return await self._one(SchemaOut, "GET", f"/schemas/{code}")

    async def create(
        self,
        body: JsonMap,
        *,
        code: str | None = None,
        name: str | None = None,
        description: str | None = None,
    ) -> SchemaOut:
        """Store a JSON Schema, taking its identity from its own keywords when none is given."""
        payload = SchemaIn(code=code, name=name, description=description, body=body)
        return await self._one(SchemaOut, "POST", "/schemas", json=request_body(payload))

    async def update(
        self,
        code: str,
        *,
        body: JsonMap | None = None,
        name: str | None = None,
        description: str | None = None,
    ) -> SchemaOut:
        """Replace a schema's body or its labels; the code is fixed."""
        payload = SchemaUpdate(name=name, description=description, body=body)
        return await self._one(SchemaOut, "PATCH", f"/schemas/{code}", json=request_body(payload))

    async def delete(self, code: str) -> None:
        """Remove a schema."""
        await self._transport.request("DELETE", f"/schemas/{code}")
